src/sources/kafka.rs

```diff
@@ -370,105 +388,415 @@ impl SourceConfig for KafkaSourceConfig {
+                    // was configured to exit. After the last such task ends, we signal the kafka
+                    // driver task to shut down the main consumer too. Note this is only used in tests.
+                    PartitionConsumerStatus::PartitionEOF if exit_eof && partition_consumers.is_empty() => {
+                        debug!("All partitions have exited or reached EOF.");
+                        let _ = eof.take().map(|e| e.send(()));
+                    },
+                    _ => {}
                 }
             },
-            message = stream.next() => match message {
-                None => break,  // WHY?
-                Some(Err(error)) => emit!(KafkaReadError { error }),
-                Some(Ok(msg)) => {
-                    emit!(KafkaBytesReceived {
-                        byte_size: msg.payload_len(),
-                        protocol: "tcp",
-                        topic: msg.topic(),
-                        partition: msg.partition(),
-                    });
+            Some(callback) = callbacks.recv() => match callback {
+                KafkaCallback::PartitionsAssigned(mut assigned_partitions, done) => {
+                    let acks = consumer.context().acknowledgements;
+                    for tp in assigned_partitions.drain(0..) {
+                        let topic = tp.0.as_str();
+                        let partition = tp.1;
+                        // It _should_ be impossible for this expect() to panic, since we receive the topic/partition pair from the rebalance callback
+                        let pq = consumer.split_partition_queue(topic, partition).expect("Failed to get partition queue: invalid topic or partition.");
+
+                        debug!("Consuming partition {}:{}.", &tp.0, tp.1);
+                        let (end_tx, handle) = partition_state.consume_partition(&mut partition_consumers, tp.clone(), Arc::clone(&consumer), pq, acks, exit_eof);
+                        abort_handles.insert(tp.clone(), handle);
+                        end_signals.insert(tp, end_tx);
+                    }
+                    // ensure this is retained until all individual queues are set up
+                    drop(done);
+                },
+                KafkaCallback::PartitionsRevoked(mut revoked_partitions, drain) => {
+                    drain_deadline = Some(Box::pin(tokio::time::sleep(max_drain_ms))).into();
+
+                    for tp in revoked_partitions.drain(0..) {
+                        if let Some(end) = end_signals.remove(&tp) {
+                            let _ = end.send(());
+                        }
+                        debug!("Revoking partition {}:{}", &tp.0, tp.1);
+                        partition_state.revoke_partition(tp);
+                    }
+
+                    if partition_state.is_drain_complete() {
+                        partition_state.clear();
+                        drop(drain);
```

Comment:
I'm still not _quite_ satisfied with how this draining state is tracked via two separate vars (a signal, and a deadline future). It sets up some pitfalls, for example this check is meant to detect when the partition tasks have already exited, and immediately proceed with the rebalance -- but the drain deadline is still set even if this check passes. We could get a spurious warning log when that deadline is reached, in this case. I'd like to package these two bits of state together in a struct or something to avoid pitfalls like this, though.

That said, I think this specific check can actually be removed, since partition task completion is now detected on another branch of this same `select!` (in a previous iteration it was detected on a different task, I believe, so this is a leftover bit of code) -- it may be the case that tasks have ended already when this check runs, but that fact will not have been registered in `partition_state` yet.